### PR TITLE
Implement Display for Arguments

### PR DESCRIPTION
**Pull request by vkgnosis**
Implement Display for our command line arguments so we don't leak secret values.

Since this allows easy custom formatting I also improved the formatting for urls and lists to be a single line.

Test Plan:
CI and manual comparison to previous way they were printed

Old discussion below:

---

We debug log our arguments in the main functions. This leaks secrets because many of them are plain strings. There are two reasonable ways to prevent this:
1. Wrap secret arguments into types with a custom Debug implementation. We do this for solver private keys.
2. Manually implement Debug/Display for the arguments.

For 1 we could introduce a generic type that wraps secret arguments with a Debug implementation that doesn't leak them.
<details>
<summary>Expand</summary>

```rust
pub struct SecretArgument<T>(pub T);

impl<T> std::fmt::Debug for SecretArgument<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("SecretArgument").finish()
    }
}

impl<T: FromStr> FromStr for SecretArgument<T> {
    type Err = <T as FromStr>::Err;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = T::from_str(s)?;
        Ok(Self(s))
    }
}
```

</details>

For 2 we would manually implement Display like is done in the code of this PR.

The downside of 2 is that it is more work to keep the Display implementation up to date. When a field is changed we need to update the impl too. This is not the case with 1. On the other hand with 1. using the wrapped arguments later is more tedious because you have to through the wrapper type like. 2 also gives more flexibility in the implementation. For example, it is annoying that we see URLs formatted over multiple lines. With 2 we can manually format every field which is why in this PR I have switched some of them to Display `{}` from Debug `{:?}`.
2 is more correct than 1. Debug implementations are supposed to be lossless and used for debugging and we are abusing the debug representation to fix our secret issues. With 2 you can still temporarily debug print the arguments while debugging which is lost with 1.

This PR has a temporary test to show the new formatting of the shared arguments:

<details>
<summary>Old</summary>

```
Arguments {
    log_filter: "warn,orderbook=debug,solver=debug,shared=debug,shared::transport::http=info",
    log_stderr_threshold: LevelFilter::ERROR,
    node_url: Url {
        scheme: "http",
        cannot_be_a_base: false,
        username: "",
        password: None,
        host: Some(
            Domain(
                "localhost",
            ),
        ),
        port: Some(
            8545,
        ),
        path: "/",
        query: None,
        fragment: None,
    },
    http_timeout: 10s,
    gas_estimators: [
        Web3,
    ],
    blocknative_api_key: None,
    base_tokens: [],
    baseline_sources: None,
    pool_cache_blocks: 10,
    pool_cache_maximum_recent_block_age: 4,
    pool_cache_maximum_retries: 5,
    pool_cache_delay_between_retries_seconds: 1s,
    block_stream_poll_interval_seconds: 5s,
    paraswap_partner: None,
    disabled_paraswap_dexs: [
        "ParaSwapPool4",
    ],
    paraswap_rate_limiter: None,
    zeroex_url: None,
    zeroex_api_key: None,
    quasimodo_uses_internal_buffers: false,
    mip_uses_internal_buffers: false,
    balancer_factories: None,
    disabled_one_inch_protocols: [
        "PMM1",
    ],
    one_inch_url: Url {
        scheme: "https",
        cannot_be_a_base: false,
        username: "",
        password: None,
        host: Some(
            Domain(
                "api.1inch.exchange",
            ),
        ),
        port: None,
        path: "/",
        query: None,
        fragment: None,
    },
    disabled_zeroex_sources: [],
    balancer_pool_deny_list: [],
    solver_competition_auth: None,
}
```

</details>

<details>
<summary>New</summary>

```
log_filter: warn,orderbook=debug,solver=debug,shared=debug,shared::transport::http=info
log_stderr_threshold: error
node_url: http://localhost:8545/
http_timeout: 10s
gas_estimators: [Web3]
blocknative_api_key: None
base_tokens: []
baseline_sources: None
pool_cache_blocks: 10
pool_cache_maximum_recent_block_age: 4
pool_cache_maximum_retries: 5
pool_cache_delay_between_retries_seconds: 1s
block_stream_poll_interval_seconds: 5s
paraswap_partner: None
disabled_paraswap_dexs: ["ParaSwapPool4"]
paraswap_rate_limiter: None
zeroex_url: None
zeroex_api_key: None
quasimodo_uses_internal_buffers: false
mip_uses_internal_buffers: false
balancer_factories: None
disabled_one_inch_protocols: ["PMM1"]
one_inch_url: https://api.1inch.exchange/
disabled_zeroex_sources: []
balancer_pool_deny_list: []
solver_competition_auth: None
```

<details>

I wanted to get this out as a draft before I convert all the arguments to the Display approach.

### Test Plan

CI, manually look at output

### CODE REVIEW

**Comment by nlordell** [comment]
I like the suggestion.

I would, however, just implement `Debug` instead of `Display`. You can do something like:
```
impl Debug for Arguments {
    fn fmt(&self, f: &Formatter) -> fmt::Result {
        f.debug_struct("Arguments")
            .field("log_filter", &format_args!("{}", self.log_filter)) // when you want `Display`
            // ...
            .field("gas_estimators", &self.gas_estimators) // when you don't
            // ...
            .finish()
    }
}
```

This also leads to not accidentally, `{:?}` debug printing and revealing secrets.

**Comment by nlordell** [comment]
Actually, you might be able to use `derivative::Derivative` for this:
```
#[derive(Derivative)]
#[derivative(Debug)]
struct Arguments {
    #[derivative(Debug(format_with="std::fmt::Display::fmt"))]
    log_filter: String,
    // ...
}
```

So that you don't accidentally forget to include new fields.

**Comment by vkgnosis** [comment]
I don't want to implement Debug because Display is more correct. Debug should show all information so we just shouldn't be using Debug for our log.

I'll look into Derivative. Not sure about it yet. Sometimes I feel doing too many things with macros is more complicated too.

**Comment by nlordell** [comment]
> I don't want to implement Debug because Display is more correct. Debug should show all information so we just shouldn't be using Debug for our log.

According to the docs:
> `Debug` should format the output in a programmer-facing, debugging context.
> `Display` is similar to `Debug`, but Display is for user-facing output, and so cannot be derived.

To me, the implementation is more "programmer-facing" than "user-facing", but this is just splitting hairs and depends on what you consider "programmer" and "user". I don't have a strong opinion either way, but would suggest removing the `derive(Debug)` implementation to avoid any mistakes that can leak secrets.

I still think we can use `debug_struct` interface as a convenient way to format it nicely (with nested indentation for example), regardless of whether this is a `Display` or `Debug` implementation:
```rust
impl Display for Arguments {
    fn fmt(&self, f: &Formatter) -> fmt::Result {
        f.debug_struct("Arguments")
            .field("log_filter", &format_args!("{}", self.log_filter)) // when you want `Display`
            // ...
            .field("gas_estimators", &self.gas_estimators) // when you don't
            // ...
            .finish()
    }
}
```

> I'll look into Derivative. Not sure about it yet. Sometimes I feel doing too many things with macros is more complicated too.

Fair enough. My one concern is that we forget to add new fields into the manual implementation (which can't happen for `derivative(Debug)`).

**Comment by vkgnosis** [comment]
I feel like Debug should not hide any information and should usually be derived. Sometimes you might need to implement it for leaf types but it doesn't make much sense to have a custom Debug implementation for aggregate structs like this. The reason we would be changing the Debug implementation is because we want the logs to not contain secrets. If I am debugging something locally I want to be able to `dbg!(arguments)` and see all the information.
When we are presenting the arguments on start up this is not the programmer debugging something. It has the goal of informing whoever runs the program of what the chosen arguments are.

**Comment by vkgnosis** [comment]
I was thinking about the indentation too but not sure we need it. We just want to print all the arguments. I'd say it is slightly preferable to not indent the shared arguments and show them as their own struct which we currently do.

**Comment by nlordell** [comment]
Just to be clear, overall I agree with the change (just have slightly different preferences on some of the details, but nothing that I feel strongly about).

**Comment by vkgnosis** [comment]
Derivative doesn't derive Display (only Debug) so I'd like to continue with the current implementation. I'm maybe 75% confident that it is the way to go and we do want to stop leaking those secrets. If it turns out to be the wrong approach going back to Debug with Derivative should be easy.

**Comment by vkgnosis** [comment]
Completed the implementation.

File: crates/orderbook/src/main.rs [hunk 1]
```diff
@@ -273,6 +274,90 @@ struct Arguments {
     pub liquidity_order_owners: Vec<H160>,
 }
 
+impl std::fmt::Display for Arguments {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.shared)?;
+        writeln!(f, "bind_address: {}", self.bind_address)?;
```
Comment (MartinquaXD) [inline crates/orderbook/src/main.rs:280]: I think to not blow up `orderbook/src/main.rs` and `solver/src/main.rs` the `Arguments` struct and its `Display` implementation should probably be moved into their own files at this point.

File: crates/orderbook/src/main.rs [hunk 2]
```diff
@@ -273,6 +274,90 @@ struct Arguments {
...
+        writeln!(f, "allowed_tokens: {:?}", self.allowed_tokens)?;
+        writeln!(f, "pool_cache_lru_size: {}", self.pool_cache_lru_size)?;
+        writeln!(f, "enable_presign_orders: {}", self.enable_presign_orders)?;
+        writeln!(
+            f,
+            "solvable_orders_max_update_age: {:?}",
+            self.solvable_orders_max_update_age
+        )?;
+        writeln!(f, "fee_discount: {}", self.fee_discount)?;
+        writeln!(f, "fee_factor: {}", self.fee_factor)?;
```
Comment (MartinquaXD) [inline crates/orderbook/src/main.rs:310]: `min_discounted_fee` is missing between those lines.

File: crates/orderbook/src/main.rs [hunk 3]
```diff
@@ -273,6 +274,90 @@ struct Arguments {
...
+        writeln!(
+            f,
+            "partner_additional_fee_factors: {:?}",
+            self.partner_additional_fee_factors
+        )?;
+        writeln!(f, "cow_fee_factors: {:?}", self.cow_fee_factors)?;
+        write!(f, "quasimodo_solver_url: ")?;
+        display_option(&self.quasimodo_solver_url, f)?;
```
Comment (MartinquaXD) [inline crates/orderbook/src/main.rs:318]: Doesn't the quasimodo URL also contain some sort of secret?

Comment (vkgnosis) [reply crates/orderbook/src/main.rs:318]: Not a secret on kubernetes. Iirc it just points to a cluster internal url.

File: crates/orderbook/src/main.rs (hunk 2 above)
Comment (vkgnosis) [reply crates/orderbook/src/main.rs:310]: Thanks.

**Review by MartinquaXD — APPROVED** [review]
Approving because I saw you already opened a PR moving the `Arguments` code into it separate files.

**Review by nlordell — APPROVED** [review]
LGTM.

Didn't verify all fields are being formatted.

File: crates/solver/src/main.rs [hunk 4]
```diff
@@ -310,6 +311,114 @@ struct Arguments {
...
+}
+
+#[test]
+fn foo() {
+    let a = Arguments::parse_from(std::iter::empty::<std::ffi::OsString>());
+    // println!("{:#?}", a);
+    // println!();
+    println!("{}", a);
+}
+
```
Comment (nlordell) [inline crates/solver/src/main.rs:421]: nit: This looks like leftover testing code that should be removed.

**Pull request merged**